org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/MakeStaticRefactoring.java

```diff
@@ -0,0 +1,599 @@
+	}
+
+	@Override
+	public RefactoringStatus checkInitialConditions(IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		fStatus= new RefactoringStatus();
+
+		SelectionInputType selectionInputType= fContextCalculator.getSelectionInputType();
+
+		if (selectionInputType == SelectionInputType.TEXT_SELECTION) {
+			fStatus.merge(InitialConditionsChecker.checkTextSelectionStart(fContextCalculator.getSelectionEditorText()));
+			fStatus.merge(InitialConditionsChecker.checkValidICompilationUnit(fContextCalculator.getSelectionICompilationUnit()));
+			if (fStatus.hasError()) {
+				return fStatus;
+			}
+
+			fContextCalculator.calculateSelectionASTNode();
+
+			fStatus.merge(InitialConditionsChecker.checkASTNodeIsValidMethod(fContextCalculator.getSelectionASTNode()));
+			if (fStatus.hasError()) {
+				return fStatus;
+			}
+
+			fContextCalculator.calculateTargetIMethodBinding();
+			fContextCalculator.calculateTargetIMethod();
+		} else {
+			//SelectionInputType is IMethod (for example when performing the refactoring form the outline menu)
+			fContextCalculator.calculateTargetIMethodBinding();
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkSourceAvailable(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkIMethodIsValid(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fContextCalculator.calculateTargetICompilationUnit();
+
+		fStatus.merge(InitialConditionsChecker.checkValidICompilationUnit(fContextCalculator.getSelectionICompilationUnit()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkMethodIsNotConstructor(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkMethodNotInLocalOrAnonymousClass(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkMethodNotStatic(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fStatus.merge(InitialConditionsChecker.checkMethodNotOverridden(fContextCalculator.getTargetIMethod()));
+		if (fStatus.hasError()) {
+			return fStatus;
+		}
+
+		fContextCalculator.calculateTargetCompilationUnit();
+		fContextCalculator.calculateMethodDeclaration();
+
+		fTargetMethod= fContextCalculator.getTargetIMethod(); //TODO remove those which dont have to necessarily be fields
+		fTargetMethodDeclaration= fContextCalculator.getTargetMethodDeclaration();
+		fTargetMethodBinding= fContextCalculator.getTargetIMethodBinding();
+
+		return fStatus;
+	}
+
+	@Override
+	public RefactoringStatus checkFinalConditions(IProgressMonitor progressMonitor) throws CoreException, OperationCanceledException {
+		modifyMethodDeclaration();
```

Comment:
I just saw that there already is some `FinalConditionChecker` that is called at some placed. Maybe you can do it the same way as for the initial conditions and the context: have a `ChangeCalculator` class and a `FinalConditionChecker` class, move everything related to the change calculation to the former class, which you pass to the final condition checker. Then this method would only call the check methods of the final condition checker, which, in turn, retrieves the required values form the change calculator. The change calculator either has the values or calculates them. Finally, the change calculator can be reused by this refactoring class to provide the final change, just by some `getOrComputeChange` method.

This would also resolve the "weirdness" of just returning a change that was calculated by the condition check on-the-fly, which you have mentioned to me before. When you clearly separate change calculation and condition checking, it becomes explicit that the change is calculates during condition checks "by accident" and that when retrieving the final change from the change calculator it simply returns a cached value (which you do not see from outside, as you could also have called the `getOrComputeChange` method without performing the condition checks before.

Reply:
@Michael5601 introduced the ChangeCalculator in d45f79f12201f4cad1062af666091a8848425a53

Reply:
Done.